sockeye/train.py

```diff
@@ -250,48 +239,128 @@ def load_or_create_vocabs(args: argparse.Namespace, resume_training: bool, outpu
-                                           bucket_width=args.bucket_width,
-                                           sequence_limit=args.limit)
+    batch_by_words = args.batch_type == C.BATCH_TYPE_WORD
+
+    either_raw_or_prepared_error_msg = "Either specify a raw training corpus with %s and %s or a preprocessed corpus " \
+                                       "with %s." % (C.TRAINING_ARG_SOURCE,
+                                                     C.TRAINING_ARG_TARGET,
+                                                     C.TRAINING_ARG_PREPARED_DATA)
+    if args.prepared_data is not None:
+        utils.check_condition(args.source is None and args.target is None, either_raw_or_prepared_error_msg)
+        if not resume_training:
+            utils.check_condition(args.source_vocab is None and args.target_vocab is None,
+                                  "You are using a prepared data folder, which is tied to a vocabulary. "
+                                  "To change it you need to rerun data preparation with a different vocabulary.")
+        train_iter, validation_iter, data_config, vocab_source, vocab_target = data_io.get_prepared_data_iters(
+            prepared_data_dir=args.prepared_data,
+            validation_source=os.path.abspath(args.validation_source),
+            validation_target=os.path.abspath(args.validation_target),
+            shared_vocab=shared_vocab,
+            batch_size=args.batch_size,
+            batch_by_words=batch_by_words,
+            batch_num_devices=batch_num_devices,
+            fill_up=args.fill_up)
+        if resume_training:
+            # resuming training. Making sure the vocabs in the model and in the prepared data match up
+            model_vocab_source = vocab.vocab_from_json(os.path.join(output_folder, C.VOCAB_SRC_NAME + C.JSON_SUFFIX))
+            model_vocab_target = vocab.vocab_from_json(os.path.join(output_folder, C.VOCAB_TRG_NAME + C.JSON_SUFFIX))
+            utils.check_condition(vocab.are_identical(vocab_source, model_vocab_source),
+                                  "Prepared data and resumed model source vocabs do not match.")
+            utils.check_condition(vocab.are_identical(vocab_target, model_vocab_target),
+                                  "Prepared data and resumed model target vocabs do not match.")
+
+        return train_iter, validation_iter, data_config, vocab_source, vocab_target
+    else:
+        utils.check_condition(args.prepared_data is None and args.source is not None and args.target is not None,
+                              either_raw_or_prepared_error_msg)
+
+        if resume_training:
+            # TODO: should we make sure it's shared?
+            # just load the vocab
+            vocab_source = vocab.vocab_from_json(os.path.join(output_folder, C.VOCAB_SRC_NAME + C.JSON_SUFFIX))
+            vocab_target = vocab.vocab_from_json(os.path.join(output_folder, C.VOCAB_TRG_NAME + C.JSON_SUFFIX))
+
+            vocab_source_path = None
```

Comment:
why not set the path here from 327 and 328?

Reply:
good question. The vocab_source_path and vocab_target_path should indicate the existing vocabs that we got before starting the initial training run. So I think what we should really do here is read the original data_config and use the vocab paths from that.